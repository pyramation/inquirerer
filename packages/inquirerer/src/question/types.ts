export interface Value {
  name: string;
  value: boolean;
}
export interface OptionValue {
  name: string;
  value: any;
}

export interface Validation {
  type?: string;
  success: boolean;
  reason?: string;
}

export interface BaseQuestion {
    name: string;
    type: string;
    default?: any;
    useDefault?: boolean;
    required?: boolean;
    message?: string;
    description?: string;
    validate?: (input: any, obj: any) => Validation | boolean;
    sanitize?: (input: any, obj: any) => any;
    pattern?: string;
  }
  
  export interface ConfirmQuestion extends BaseQuestion {
    type: 'confirm';
    default?: boolean;  // Defaults are typically boolean for confirm types
  }
  
  export interface AutocompleteQuestion extends BaseQuestion {
    type: 'autocomplete';
    options: (string|OptionValue)[];
    maxDisplayLines?: number;
    returnFullResults?: boolean;
    allowCustomOptions?: boolean;
  }
  
  export interface CheckboxQuestion extends BaseQuestion {
    type: 'checkbox';
    options: (string|OptionValue)[];
    maxDisplayLines?: number;
    returnFullResults?: boolean;
    default?: string[];
  }
  
  export interface TextQuestion extends BaseQuestion {
    type: 'text';
    default?: string;
  }

  export interface NumberQuestion extends BaseQuestion {
    type: 'number';
    default?: number;
  }
  
  export type Question = ConfirmQuestion | AutocompleteQuestion | CheckboxQuestion | TextQuestion | NumberQuestion;