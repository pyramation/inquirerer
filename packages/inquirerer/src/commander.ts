import deepmerge from 'deepmerge';
import minimist, { Opts, ParsedArgs } from 'minimist';
import { Readable,Writable } from 'stream';

import { Inquirerer } from './prompt';
import { getVersion } from './utils';

// Define the type for the command handler function
export type CommandHandler = (argv: ParsedArgs, prompter: Inquirerer, options: CLIOptions) => void;

export interface CLIOptions {
  noTty: boolean;
  input: Readable;
  output: Writable;
  minimistOpts: Opts;
  version: string;
}

export const defaultCLIOptions: CLIOptions = {
  version: `inquirerer@${getVersion()}`,
  noTty: false,
  input: process.stdin,
  output: process.stdout,
  minimistOpts: {
    alias: {
      v: 'version'
    }
  }
};

export class CLI {
  private argv: ParsedArgs;
  private prompter: Inquirerer;
  private commandHandler: CommandHandler;
  private options: CLIOptions;

  constructor(
    commandHandler: CommandHandler,
    options: Partial<CLIOptions>,
    argv?: any
  ) {
    const { input, output, ...optionsWithoutIO } = options;
    const { input: defaultInput, output: defaultOutput, ...defaultOptionsWithoutIO } = defaultCLIOptions;
    const mergedOptions: Partial<CLIOptions> = deepmerge(defaultOptionsWithoutIO, optionsWithoutIO);
    mergedOptions.input = input || defaultInput;
    mergedOptions.output = output || defaultOutput;
    this.options = mergedOptions as CLIOptions;

    this.argv = argv ? argv : minimist(process.argv.slice(2), this.options.minimistOpts);

    this.prompter = new Inquirerer({
      noTty: this.options.noTty,
      input: this.options.input,
      output: this.options.output
    });
    this.commandHandler = commandHandler;

  }

  public async run(): Promise<void> {
    if (!('tty' in this.argv)) {
      this.argv.tty = true;
    }

    if (this.argv.version) {
      console.log(this.options.version);
      process.exit(0);
    }

    const args = await this.commandHandler(this.argv, this.prompter, this.options);
    this.prompter.close();
    return args;
  }
}

export default CLI;
