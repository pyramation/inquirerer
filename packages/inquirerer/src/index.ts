export * from './prompt';
export * from './question';