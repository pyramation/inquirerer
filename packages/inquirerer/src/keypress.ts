type KeyHandler = () => void;

export const KEY_CODES = {
  UP_ARROW: '\u001b[A',
  DOWN_ARROW: '\u001b[B',
  RIGHT_ARROW: '\u001b[C',
  LEFT_ARROW: '\u001b[D',
  ENTER: '\r',
  SPACE: ' ',
  CTRL_C: '\u0003',
  BACKSPACE: '\x7f',  // Commonly used BACKSPACE key in Unix-like systems
  BACKSPACE_LEGACY: '\x08'  // For compatibility with some systems
};

export class TerminalKeypress {
  private listeners: Record<string, KeyHandler[]> = {};
  private active: boolean = true;


  constructor() {
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.setEncoding('utf8');
    this.setupListeners();
  }


  private setupListeners(): void {
    process.stdin.on('data', (key: string) => {
      if (!this.active) return;
      const handlers = this.listeners[key];
      handlers?.forEach(handler => handler());
      if (key === KEY_CODES.CTRL_C) { // Ctrl+C
        process.exit();
      }
    });
  }

  on(key: string, callback: KeyHandler): void {
    if (!this.listeners[key]) {
      this.listeners[key] = [];
    }
    this.listeners[key].push(callback);
  }

  off(key: string, callback: KeyHandler): void {
    if (this.listeners[key]) {
      const index = this.listeners[key].indexOf(callback);
      if (index !== -1) {
        this.listeners[key].splice(index, 1);
      }
    }
  }

  pause(): void {
    this.active = false;
  }

  resume(): void {
    this.active = true;
  }

  destroy(): void {
    process.stdin.setRawMode(false);
    process.stdin.pause();
    process.stdin.removeAllListeners('data');
  }
}
