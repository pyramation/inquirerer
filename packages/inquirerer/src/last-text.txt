{
  "question": {
    "type": "text",
    "name": "last",
    "required": true,
    "message": "Enter your last name"
  },
  "obj": {
    "fruitSearch": "Apricot",
    "first": "adsf",
    "last": "asdf"
  }
}