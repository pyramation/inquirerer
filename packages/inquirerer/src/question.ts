export interface Question {
  name: string;
  type?: string; // This can be used for further customizations like validating input based on type

  options?: string[],

  maxDisplayLines?: number;  // Optional parameter to limit the number of visible options

  returnFullResults?: boolean;
  allowCustomOptions?: boolean;

  required?: boolean;
}
