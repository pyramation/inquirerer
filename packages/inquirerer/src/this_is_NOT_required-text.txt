{
  "question": {
    "name": "this_is_NOT_required",
    "type": "checkbox",
    "maxDisplayLines": 5,
    "options": [
      "Apple",
      "Apricot",
      "Avocado",
      "Banana",
      "Blackberry",
      "Blueberry",
      "Boysenberry",
      "Cherry",
      "Clementine",
      "Coconut",
      "Cranberry",
      "Date",
      "Durian",
      "Elderberry",
      "Fig",
      "Grape",
      "Grapefruit",
      "Guava",
      "Honeydew",
      "Kiwi",
      "Kumquat",
      "Lemon",
      "Lime",
      "Lychee",
      "Mango",
      "Melon",
      "Mulberry",
      "Nectarine",
      "Orange",
      "Papaya",
      "Peach",
      "Pear",
      "Persimmon",
      "Pineapple",
      "Plum",
      "Pomegranate",
      "Pomelo",
      "Quince",
      "Raspberry",
      "Redcurrant",
      "Strawberry",
      "Starfruit",
      "Tangerine",
      "Ugli Fruit",
      "Vanilla",
      "Watermelon",
      "Xigua (Chinese Watermelon)",
      "Yellow Plum",
      "Zucchini"
    ],
    "required": true
  },
  "obj": {
    "fruitSearch": "Apple",
    "first": null,
    "last": "sdf",
    "autocomp": "Apple",
    "this_is_NOT_required": []
  }
}