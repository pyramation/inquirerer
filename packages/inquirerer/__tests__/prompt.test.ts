import readline from 'readline';
import { Readable, Writable, Transform } from 'stream';


import { Inquirerer } from '../src';
import { Question } from '../src/question';

jest.mock('readline');

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Inquirerer', () => {
  let mockWrite: jest.Mock;
  let mockInput: Readable;
  let mockOutput: Writable;
  let questionHandlers: Array<(input: string) => void> = [];
  let currentQuestionIndex: number = 0;
  let transformStream: Transform;

  let writeResults: string[];
  let transformResults: string[];

  let inputQueue: Array<{ type: 'key' | 'read', value: string }> = [];
  let currentInputIndex: number = 0;

  function setupReadlineMock() {
    readline.createInterface = jest.fn().mockReturnValue({
      question: (questionText: string, cb: (input: string) => void) => {
        // Process the queued inputs when question is called
        const nextInput = inputQueue[currentInputIndex++];
        if (nextInput && nextInput.type === 'read') {
          setTimeout(() => cb(nextInput.value), 350); // Simulate readline delay
        }
      },
      close: jest.fn(),
    });
  }

  function enqueueInputResponse(input: { type: 'key' | 'read', value: string }) {
    if (input.type === 'key') {
  // Push key events directly to mockInput
  // @ts-ignore
      setTimeout(() => mockInput.push(input.value), 350);
    } else {
      // Queue readline responses to be handled by the readline mock
      inputQueue.push(input);
    }
  }

  const KEY_SEQUENCES = {
    ENTER: '\u000d',     // Carriage return
    UP_ARROW: '\u001b[A', // ANSI escape sequence for the up arrow
    DOWN_ARROW: '\u001b[B',
    SPACE: ' '
  };

  beforeEach(() => {
    mockWrite = jest.fn();
    currentQuestionIndex = 0;
    questionHandlers = [];
    writeResults = [];
    transformResults = [];


    mockInput = new Readable({
      read(size) { }
    });
    // @ts-ignore
    mockInput.setRawMode = jest.fn();  // Mock TTY-specific method if needed

    mockOutput = new Writable({
      write: (chunk, encoding, callback) => {
        const str = chunk.toString();
        writeResults.push(str);
        mockWrite(str);
        callback();
      }
    });

    // Create the transform stream to log and pass through data
    transformStream = new Transform({
      transform(chunk, encoding, callback) {
        const data = chunk.toString();
        transformResults.push(data);
        this.push(chunk); // Pass the data through
        callback();
      }
    });

    setupReadlineMock();
    // Pipe the transform stream to the mock output to intercept writes
    //  transformStream.pipe(mockOutput);

    // mockOutput.pipe(transformStream);
    mockInput.pipe(transformStream);

    // mockInput.pipe(transformStream).pipe(mockOutput);

  });

  it('prompts user and correctly processes delayed input', async () => {
    enqueueInputResponse({ type: 'read', value: 'user input' });

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [{ name: 'username', type: 'text' }];
    const initialParams = {};

    const expectedResult = { username: 'user input' };

    const result = await prompter.prompt(initialParams, questions);

    expect(result).toEqual(expectedResult);
    // expect(mockWrite).toHaveBeenCalledWith(expect.stringContaining('username:\n> '));
    expect(writeResults).toMatchSnapshot();
    expect(transformResults).toMatchSnapshot();
  });

  it('handles multiple questions', async () => {
    enqueueInputResponse({ type: 'read', value: 'first question answer' });
    enqueueInputResponse({ type: 'read', value: 'second question answer' });

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [
      { name: 'firstQuestion', type: 'text' },
      { name: 'secondQuestion', type: 'text' }
    ];
    const initialParams = {};

    const result = await prompter.prompt(initialParams, questions);

    expect(result).toEqual({
      firstQuestion: 'first question answer',
      secondQuestion: 'second question answer'
    });
    expect(writeResults).toMatchSnapshot();
    expect(transformResults).toMatchSnapshot();
  });

  xit('handles autocomplete selection with key events', async () => {
    // Simulate typing 'fir', navigating down to the second option, and selecting it
    // enqueueInputResponse({ type: 'read', value: 'fir' });
    // enqueueInputResponse({ type: 'key', value: 'DOWN_ARROW' });

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [{
      name: 'autocompleteField',
      type: 'autocomplete',
      options: ['first option', 'firry second option', 'firry third option']
    }];
    const initialParams = {};

    enqueueInputResponse({ type: 'key', value: 'f' });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER });

    const result = await prompter.prompt(initialParams, questions);

    // Expected to select the second option
    expect(result).toEqual({ autocompleteField: 'second option' });
    expect(writeResults).toMatchSnapshot();
    expect(transformResults).toMatchSnapshot();
  });

  it('handles combined key events and readline inputs', async () => {

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [
      { name: 'firstQuestion', type: 'text' },
      { name: 'secondQuestion', type: 'text' }
    ];
    const initialParams = {};

    enqueueInputResponse({ type: 'read', value: 'first question answer' });
    // enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    // enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER }); // Simulate Enter
    enqueueInputResponse({ type: 'read', value: 'second question answer' });

    const result = await prompter.prompt(initialParams, questions);



    expect(result).toEqual({
      firstQuestion: 'first question answer',
      secondQuestion: 'second question answer'
    });
  });

  it('checkbox', async () => {

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [
      {
        name: 'checkbox', type: 'checkbox', options: [
          'a',
          'b',
          'c'
        ]
      }
    ];
    const initialParams = {};

    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.SPACE }); // Space to select
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.SPACE }); // Space to select
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER }); // Simulate Enter

    const result = await prompter.prompt(initialParams, questions);

    expect(result).toMatchSnapshot();
  });

  it('checkbox w/options', async () => {

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [
      { 
        name: 'checkbox',
        type: 'checkbox',
        options: [
          'a',
          'b',
          'c'
        ],
        returnFullResults: true
      }
    ];
    const initialParams = {};

    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.SPACE }); // Space to select
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.SPACE }); // Space to select
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER }); // Simulate Enter

    const result = await prompter.prompt(initialParams, questions);

    expect(result).toMatchSnapshot();
  });

  it('handles readline inputs', async () => {

    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [
      { name: 'firstQuestion', type: 'text' },
      { name: 'secondQuestion', type: 'text' },
      { name: 'youGotIt', type: 'confirm' },
      { name: 'andGotIt', type: 'confirm' },
    ];
    const initialParams = {};

    enqueueInputResponse({ type: 'read', value: 'first question answer' });
    // enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW }); // Simulate Down Arrow
    // enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER }); // Simulate Enter
    enqueueInputResponse({ type: 'read', value: 'second question answer' });
    enqueueInputResponse({ type: 'read', value: 'Y' });
    enqueueInputResponse({ type: 'read', value: 'N' });

    const result = await prompter.prompt(initialParams, questions);



    expect(result).toEqual({
      firstQuestion: 'first question answer',
      secondQuestion: 'second question answer',
      youGotIt: true,
      andGotIt: false
    });
  });

});
