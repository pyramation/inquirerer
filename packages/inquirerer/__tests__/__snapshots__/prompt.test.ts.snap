// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inquirerer checkbox 1`] = `
{
  "checkbox": [
    {
      "name": "b",
      "value": true,
    },
    {
      "name": "c",
      "value": true,
    },
  ],
}
`;

exports[`Inquirerer checkbox w/options 1`] = `
{
  "checkbox": [
    {
      "name": "a",
      "value": false,
    },
    {
      "name": "b",
      "value": true,
    },
    {
      "name": "c",
      "value": true,
    },
  ],
}
`;

exports[`Inquirerer handles autocomplete selection with key events 1`] = `
[
  "c",
  "autocompleteField:
> 
",
  "[34m> first option[39m
",
  "  second option
",
  "  third option
",
  "c",
  "autocompleteField:
> 
",
  "  first option
",
  "[34m> second option[39m
",
  "  third option
",
]
`;

exports[`Inquirerer handles autocomplete selection with key events 2`] = `
[
  "[B",
  "
",
]
`;

exports[`Inquirerer handles multiple questions 1`] = `
[
  "c",
  "c",
]
`;

exports[`Inquirerer handles multiple questions 2`] = `[]`;

exports[`Inquirerer prompts user and correctly processes delayed input 1`] = `
[
  "c",
]
`;

exports[`Inquirerer prompts user and correctly processes delayed input 2`] = `[]`;
